### PR TITLE
Presets to Patch, Patches in db, drag & drop Entities (and Components); resolves #1766

### LINKED ISSUE #1766: Drag & Drop Behaviors and Materials onto 3D Objects

### PR DESCRIPTION
**Pull request by kschzt**
- Improve auto-layout of entities in root patch
- Rename Presets to Patches
- Import Patches to DB
- Read system Patches list from db 
- Drag & drop of entities into scene
- Open selected entity patch in program mode
- Select entity in scene when selecting in program mode
- Use entity.plugin instead of graph.plugin on patches that have a single `object3d` output
- Improve GraphAnalyser
- Move the activation of a graph from TreeView to editor
- Always drop entities into root graph
- migration scripts
-- move system patches to db
-- fix patch paths in db
-- change Object patches to use Entity or EntityComponent

Will clean up the commit log in the end.

### CODE REVIEW

File: services/assetService.js [hunk 1]
```diff
@@ -113,40 +107,34 @@ AssetService.prototype.findByPath = function(path) {
...
-			asset.previewUrlSmall = data.previewUrlSmall
+		if (!asset) {
+			asset = new that._model(data)
+			asset._creator = user.id
+		}
 
-		if (data.previewUrlLarge)
-			asset.previewUrlLarge = data.previewUrlLarge
+		// update model with stuff from data
+		_.assign(asset, data)
```
Comment (gmarinov) [inline services/assetService.js]: so passing data with existing but empty keys will now overwrite the asset? double-checking this is desired.

File: services/assetService.js [hunk 2]
```diff
@@ -1,32 +1,31 @@
...
-AssetService.prototype.find = function(q)
-{
-	var dfd = when.defer();
+AssetService.prototype.find = function(q) {
+	var dfd = when.defer()
 	this._model
 		.find(q)
 		.sort('-updatedAt')
-		.populate('_creator')
+		// .populate('_creator')
```
Comment (gmarinov) [inline services/assetService.js]: double-check this?

File: models/patch.js [hunk 3]
```diff
@@ -0,0 +1,13 @@
...
+var assetHelper = require('./asset-helper');
+var _ = require('lodash');
+
+var patchSchema = new mongoose.Schema(_.assign({
+	category: { type: String },
+	name: { type: String, required: true },
+	type: { type: String, required: true },
+	stat: { type: Schema.Types.Mixed, required: true }
+}, assetHelper.schema), {});
+
```
Comment (gmarinov) [inline models/patch.js]: is it a good time to add icon: { ... } ?

File: lib/gridfs-storage.js [hunk 4]
```diff
@@ -86,6 +84,10 @@ GridFsStorage.prototype.createWriteStream = function(path, mimetype) {
 }
 
 GridFsStorage.prototype.writeString = function(path, stringData, encoding) {
+	if (!stringData) {
+		console.trace('writeString got no stringData', path)
+	}
```
Comment (gmarinov) [inline lib/gridfs-storage.js]: should this also return early here?

File: common/graphAnalyser.js [hunk 5]
```diff
@@ -26,10 +29,44 @@ GraphAnalyser.prototype.parseAssets = function(graph) {
...
+		type: 'patch'
 	}
 
-	function findInGraph(subgraph) {
+	function readRoot(graph) {
+		if (!graph.nodes || !graph.nodes.length)
+			return console.error('No nodes to GraphAnalyser.parseAssets')
+
+		var patchNode = graph.nodes[0]
+		var patchNodeId = patchNode.plugin
```
Comment (gmarinov) [inline common/graphAnalyser.js]: this doesn't look right to me? `patchNode.id` ?

File: browser/scripts/worldEditor/worldEditor.js [hunk 6]
```diff
@@ -408,106 +364,270 @@ WorldEditor.prototype.selectMeshAndDependencies = function(meshNode, sceneNode,
...
-
-	// if alt is pressed, we try to select the single object we click
-	// otherwise we select the topmost group the clicked object is in
-	var selectSingleObject = E2.ui.flags.pressedAlt
+WorldEditor.prototype.onPatchDropped = function(patchMeta, json, targetObject3d) {
+	var sceneNode = this.currentGroup || this.scene.backReference.parentNode
+	var leftFromSceneNode = 300
 
-	var isEditor = this.isActive()
+	console.debug('onPatchDropped', patchMeta, !!targetObject3d)
```
Comment (gmarinov) [inline browser/scripts/worldEditor/worldEditor.js]: still need this?

File: browser/scripts/graphApi.js [hunk 7]
```diff
@@ -5,6 +5,37 @@ function GraphApi(undoManager) {
...
+			src_nuid: patchNode.uid,
+			dst_nuid: targetNode.uid,
+			src_slot: output.index,
+			src_dyn: true,
+			dst_slot: output.name,
+			dst_dyn: false
+		})
+
+		// figure out how to mix values on busy slots
+		// instead of replacing the connections
```
Comment (gmarinov) [inline browser/scripts/graphApi.js]: is this waffled? looks like a todo

File: browser/scripts/graph.js [hunk 8]
```diff
@@ -20,6 +20,19 @@ function Graph(core, parent_graph, uid) {
...
 
+Graph.prototype.isEntityPatch = function() {
+	if (!this.plugin)
+		return false
+
+	if (this.plugin.id === 'entity')
+		return true
+
+	var node = this.plugin.node
+	node.isEntityPatch()
```
Comment (gmarinov) [inline browser/scripts/graph.js]: double-check this? looks unused.

File: browser/scripts/application.js [hunk 9]
```diff
@@ -1147,10 +1181,10 @@ Application.prototype.getNodeBoundingBox = function(node) {
 
 	// default width / height = 100 / 20 if the graph is not visible
```
Comment (gmarinov) [inline browser/scripts/application.js]: probably worth removing the digits from comment

File: browser/scripts/application.js [hunk 10]
```diff
@@ -936,31 +925,38 @@ Application.prototype.selectionToObject = function(nodes, conns, sx, sy) {
...
-	return JSON.stringify(this.selectionToObject(nodes, conns, sx, sy))
+Application.prototype.stringifyNodesAndConnections = function(nodes, conns, sx, sy) {
+	return JSON.stringify(
+		this.selectionToObject(nodes, conns, sx, sy)
+	)
 }
 
 Application.prototype.onDelete = function(e) {
-	if (!this.selectedNodes.length)
+	if (E2.ui.isInProgramMode() && !this.selectedNodes.length)
```
Comment (gmarinov) [inline browser/scripts/application.js]: the flow here is a bit confusing. can haz some build/program clarity?

File: browser/scripts/application.js [hunk 11]
```diff
@@ -910,15 +900,14 @@ Application.prototype.selectionToObject = function(nodes, conns, sx, sy) {
 
 		var b = [p.left, p.top, p.left + width, p.top + height];
 
-		if(dom)
-			n = n.serialise();
+		// n = n.serialise();
```
Comment (gmarinov) [inline browser/scripts/application.js]: still need this?

File: browser/scripts/abstractPlugins/abstractObjectGazePlugin.js [hunk 12]
```diff
@@ -9,7 +9,7 @@
 				return
 
 			subgraph.nodes.map(function(node) {
-				if (node.plugin.id === 'graph')
+				if (E2.GRAPH_NODES.indexOf(node.plugin.id) > -1)
```
Comment (gmarinov) [inline browser/scripts/abstractPlugins/abstractObjectGazePlugin.js]: found this pattern everywhere but over here specifically a plugin knows something about E2.GRAPH_NODES. justsaying..

File: browser/plugins/plugins.json [hunk 13]
```diff
@@ -146,18 +145,25 @@
         "Object Gaze Clicked": "on_three_object_gaze_clicked"
     },
 
-
     "STATE AND STRUCTURE": {
         "Annotation": "annotation",
+
+        "Is In VR": "is_in_vr",
```
Comment (gmarinov) [inline browser/plugins/plugins.json]: 'Is in VR' but also our capitalisation overall isn't great ('render To texture' above) so maybe leave for another pass.

File: browser/scripts/abstractPlugins/abstractGraphPlugin.js [hunk 14]
```diff
@@ -0,0 +1,87 @@
...
+	this.input_slots = []
+	this.output_slots = []
+	
+	this.state = {
+		always_update:  true,
+		input_sids:     {},
+		output_sids:    {}
+	}
+		
+	this.is_reset = true
```
Comment (gmarinov) [inline browser/scripts/abstractPlugins/abstractGraphPlugin.js]: I couldn't find anywhere .is_reset is used

File: services/assetService.js (hunk 1 above)
Comment (kschzt) [inline services/assetService.js]: @gmarinov well yeah, basically, whatever's input we want to pass through to the model.

File: services/assetService.js (hunk 2 above)
Comment (kschzt) [inline services/assetService.js]: yep, we don't want to *always* populate the creator - I'll remove the comment

File: models/patch.js (hunk 3 above)
Comment (kschzt) [inline models/patch.js]: @gmarinov nope, that's out of the scope of this pull request..

File: common/graphAnalyser.js (hunk 5 above)
Comment (kschzt) [inline common/graphAnalyser.js]: @gmarinov it's called `plugin` in the serialization :)

File: browser/scripts/abstractPlugins/abstractObjectGazePlugin.js (hunk 12 above)
Comment (kschzt) [inline browser/scripts/abstractPlugins/abstractObjectGazePlugin.js]: @gmarinov it's better than checking for 'graph' ;) which broke entity, entity_component and loop ...

**Comment by gmarinov** [comment]
lgtm

**Comment by kkananen** [comment]
lgtm

**Pull request merged**